File: Notes/app/src/main/java/com/serhii/apps/notes/ui/view_model/LoginViewModel.java
package com.serhii.apps.notes.ui.view_model;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.serhii.apps.notes.control.EventService;
import com.serhii.apps.notes.control.base.IAuthorizeService;
import com.serhii.apps.notes.ui.data_model.AuthModel;

/**
 *  View model for managing authentication data and User's actions.
 */

public class LoginViewModel extends ViewModel {

    private final IAuthorizeService authorizeService = EventService.getInstance();
    private final MutableLiveData<Boolean> showRegistrationUISetFlag = new MutableLiveData<>(false);

    public void setAuthValue(AuthModel newValue) {
        // If data is set then perform corresponding action
        if (newValue != null) {
            switch (newValue.getAuthType()) {
                case AUTH_UNLOCK:
                case AUTH_PASSWORD_LOGIN:
                    // Proceed with Login
                    authorizeService.onPasswordLogin(newValue);
                    break;
                case AUTH_REGISTRATION:
                    // Proceed with User's registration
                    authorizeService.onRegistration(newValue);
                    break;
                case AUTH_BIOMETRIC_LOGIN:
                    // Proceed with Login
                    authorizeService.onBiometricLogin();
                    break;
            }
        }
    }

    public IAuthorizeService getAuthorizeService() { return authorizeService; }

    public void showRegistrationUI() {
        // Notify observers that action should be performed
        showRegistrationUISetFlag.setValue(true);
        // Reset flag to notify that activity doesn't need to perform any action
        showRegistrationUISetFlag.setValue(false);
    }

    public LiveData<Boolean> getShowRegistrationUISetFlag() {
        return showRegistrationUISetFlag;
    }

}

File: Notes/app/src/main/java/com/serhii/apps/notes/ui/view_model/LoginViewModel.kt
/*
 * Copyright 2023. Happy coding ! :)
 * Author: Serhii Butryk
 */
package com.serhii.apps.notes.ui.view_model

import android.content.Context
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.serhii.apps.notes.common.App
import com.serhii.apps.notes.control.EventService
import com.serhii.apps.notes.control.auth.base.IEventService
import com.serhii.apps.notes.control.auth.types.AuthorizeType
import com.serhii.apps.notes.ui.data_model.AuthModel
import com.serhii.core.security.BiometricAuthenticator
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.launch

/**
 *  View model for managing UI state of AuthorizationActivity
 */
class LoginViewModel : ViewModel() {

    val showRegistrationUI = MutableLiveData(false)

    private val authorizeService: IEventService = EventService

    fun proceedWithAuth(context: Context, authModel: AuthModel) {
        // Launch a new coroutine to handle this request
        viewModelScope.launch(App.BACKGROUND_DISPATCHER
                + CoroutineName("AuthRequest")) {
            when (authModel.authType) {
                AuthorizeType.AUTH_UNLOCK, AuthorizeType.AUTH_PASSWORD_LOGIN ->
                    authorizeService.onPasswordLogin(context, authModel)
                AuthorizeType.AUTH_REGISTRATION -> {
                    throw RuntimeException("Not a right method")
                }
                AuthorizeType.AUTH_BIOMETRIC_LOGIN ->
                    authorizeService.onBiometricLogin(context, authModel)
                else -> {
                    throw RuntimeException("Unknown auth type")
                }
            }
        }
    }

    fun proceedWithRegistration(authModel: AuthModel, biometricAuthenticator: BiometricAuthenticator?, fragmentActivity: FragmentActivity) {
        // Launch a new coroutine to handle this request
        val context = App.BACKGROUND_DISPATCHER + CoroutineName("RegistrationRequest")
        viewModelScope.launch(context) {
            authorizeService.onRegistration(authModel, biometricAuthenticator, fragmentActivity, this)
        }
    }

    fun requestRegistrationUI() {
        // Notify observers that we need to show an activity
        showRegistrationUI.value = true
        // Reset flag
        showRegistrationUI.value = false
    }
}